# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [8]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [9]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [10]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [11]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [12]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [13]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [14]:
df.nunique()

Season     32
Daynum    133
Wteam     364
Wscore    130
Lteam     364
Lscore    117
Wloc        3
Numot       7
dtype: int64

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [15]:
df["Wteam"].value_counts()

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [16]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [18]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [19]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [20]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [21]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [22]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [23]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [24]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [25]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [26]:
df.sort_values("Season")

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
2484      1985      99   1275      73   1132      63    H      0
2485      1985      99   1306      62   1171      59    H      0
2486      1985      99   1318      99   1145      91    H      0
2487      1985      99   1337      75   1109      56    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
141706    2016      50   1300      57   1183      52    N      0
141705    2016      50   1298      62   1202      49    H      0
141704    2016      50   1295      68   1410      50    N      0
141702    2016      50   1277      99   1324      93    N      1
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

In [27]:
df.sort_values("Wteam")

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      62   1101      87   1146      70    H      0
143283    2016      89   1101      80   1249      71    H      0
143476    2016      92   1101      79   1223      72    H      0
141972    2016      61   1101      62   1146      61    A      0
...        ...     ...    ...     ...    ...     ...  ...    ...
56143     1998     125   1464      75   1282      59    N      0
94008     2007      41   1464      68   1454      65    A      0
41936     1995      84   1464      67   1282      52    H      0
110589    2010      68   1464      71   1260      67    H      0
46638     1996      98   1464      73   1442      48    H      0

[145289 rows x 8 columns]

In [28]:
df.sort_values("Wscore")

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336     2001     126   1206      35   1423      33    N      0
5364      1986      77   1229      35   1166      34    A      0
20246     1990      47   1374      36   1426      31    H      0
118590    2011     130   1336      36   1458      33    N      0
...        ...     ...    ...     ...    ...     ...  ...    ...
20022     1990      40   1116     166   1109     101    H      0
24341     1991      47   1328     172   1258     112    H      0
19653     1990      30   1328     173   1109     101    H      0
17867     1989      92   1258     181   1109     150    H      0
24970     1991      68   1258     186   1109     140    H      0

[145289 rows x 8 columns]

In [29]:
df.iloc[100:201][["Season", "Wteam", "Wscore"]]

Season  Wteam  Wscore
100    1985   1409      86
101    1985   1444      90
102    1985   1451      89
103    1985   1139      78
104    1985   1154      78
..      ...    ...     ...
196    1985   1203      83
197    1985   1222      73
198    1985   1227      76
199    1985   1228      73
200    1985   1229      54

[101 rows x 3 columns]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [30]:
max_score_index=df['Wscore'].idxmax()

In [31]:
a=df.loc[max_score_index,"Season"]
b=df.loc[max_score_index, "Wteam"]

In [32]:
print("Season:", a)
print("Wteam:", b)

Season: 1991
Wteam: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [33]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [34]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [49]:
import pandas as pd

df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

ScoreDifference = df['ScoreDifference']

In [50]:
average_difference=df["ScoreDifference"].mean()
average_difference

12.103311331208833

In [51]:
filtered_df = df[df["ScoreDifference"]>average_difference]
filtered_df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [52]:
max_diff_index = df['ScoreDifference'].idxmax()
season = df.loc[max_diff_index, 'Season']
winning_team = df.loc[max_diff_index, 'Wteam']
losing_team = df.loc[max_diff_index, 'Lteam']
print("Season:", season)
print("Winning Team:", winning_team)
print("Losing Team:", losing_team)

Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [53]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [54]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [55]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [56]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [57]:
df.groupby("Season")["Wscore"].mean().head()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [58]:
df.groupby("Wteam")["Wscore"].max()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [59]:
X = df.groupby(["Wteam", "Season"])["Wscore"].mean()

In [60]:
X.sort_values()

Wteam  Season
1126   2007       52.571429
1136   2010       53.000000
1109   1985       54.000000
1311   2003       55.000000
1185   2013       55.076923
                    ...    
1258   1991      108.428571
       1988      111.750000
1380   1991      112.312500
1258   1989      117.315789
       1990      126.347826
Name: Wscore, Length: 10172, dtype: float64

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [61]:
df.groupby(["Wteam", "Season"])["Wscore"].count()

Wteam  Season
1101   2014       2
       2015       7
       2016       9
1102   1985       5
       1986       6
                 ..
1464   2012      14
       2013      14
       2014      12
       2015       8
       2016       9
Name: Wscore, Length: 10172, dtype: int64

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [62]:
season_wins = df.groupby(["Season", "Wteam"]). size().reset_index(name="Wscore")
team_with_most_wins = season_wins.groupby("Season")["Wscore"].idxmax()
team_most_wins_per_season = season_wins.loc[team_with_most_wins]

for index, row in team_most_wins_per_season.iterrows():
    season = row["Season"]
    team = row["Wteam"]
    wins = row["Wscore"]
    print(f"Season: {season}, Team with most wins: {team}, Wins: {wins}")

Season: 1985, Team with most wins: 1385, Wins: 27
Season: 1986, Team with most wins: 1181, Wins: 32
Season: 1987, Team with most wins: 1424, Wins: 33
Season: 1988, Team with most wins: 1112, Wins: 31
Season: 1989, Team with most wins: 1328, Wins: 28
Season: 1990, Team with most wins: 1247, Wins: 29
Season: 1991, Team with most wins: 1116, Wins: 30
Season: 1992, Team with most wins: 1181, Wins: 28
Season: 1993, Team with most wins: 1231, Wins: 28
Season: 1994, Team with most wins: 1163, Wins: 27
Season: 1995, Team with most wins: 1116, Wins: 26
Season: 1996, Team with most wins: 1269, Wins: 31
Season: 1997, Team with most wins: 1242, Wins: 31
Season: 1998, Team with most wins: 1242, Wins: 33
Season: 1999, Team with most wins: 1181, Wins: 32
Season: 2000, Team with most wins: 1409, Wins: 29
Season: 2001, Team with most wins: 1181, Wins: 29
Season: 2002, Team with most wins: 1153, Wins: 30
Season: 2003, Team with most wins: 1166, Wins: 29
Season: 2004, Team with most wins: 1390, Wins: 29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [63]:
df.groupby("Lteam")["Lscore"].mean()

Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [64]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [68]:
high_scoring_game = []

for index, row in df.iterrows():
    if row['Wscore'] > 100:
        high_scoring_game.append('Yes')
    else:
        high_scoring_game.append('No')
        
df['HighScoringGame'] = high_scoring_game

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [70]:
games_played = {}

for index, row in df.iterrows():
    
    # Get the winning team and update the count
    winning_team = row['Wteam']
    games_played[winning_team] = games_played.get(winning_team, 0) + 1
    
    # Get the losing team and update the count
    losing_team = row['Lteam']
    games_played[losing_team] = games_played.get(losing_team, 0) + 1

In [71]:
# Get the number of games played by each team
for team, count in games_played.items():
    print(f"Team: {team}, Games Played: {count}")

Team: 1228, Games Played: 992
Team: 1328, Games Played: 968
Team: 1106, Games Played: 855
Team: 1354, Games Played: 906
Team: 1112, Games Played: 981
Team: 1223, Games Played: 363
Team: 1165, Games Played: 833
Team: 1432, Games Played: 69
Team: 1192, Games Played: 908
Team: 1447, Games Played: 903
Team: 1218, Games Played: 931
Team: 1337, Games Played: 922
Team: 1226, Games Played: 847
Team: 1242, Games Played: 993
Team: 1268, Games Played: 969
Team: 1260, Games Played: 914
Team: 1133, Games Played: 949
Team: 1305, Games Played: 922
Team: 1424, Games Played: 974
Team: 1307, Games Played: 969
Team: 1288, Games Played: 925
Team: 1344, Games Played: 951
Team: 1438, Games Played: 952
Team: 1374, Games Played: 916
Team: 1411, Games Played: 903
Team: 1412, Games Played: 962
Team: 1397, Games Played: 963
Team: 1417, Games Played: 966
Team: 1225, Games Played: 880
Team: 1116, Games Played: 980
Team: 1368, Games Played: 808
Team: 1120, Games Played: 936
Team: 1391, Games Played: 879
Team: 1135,

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [72]:
highest_score_diff_per_season = {}

for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']
    
    # We are going to check if the season already exists in the dictionary
    if season in highest_score_diff_per_season:
        
    # If the current score difference is higher, update the value in the dictionary
        if score_diff > highest_score_diff_per_season[season]['score_diff']:
            highest_score_diff_per_season[season] = {
                'score_diff': score_diff,
                'game': index
            }
    else:
        # If the season is not in the dictionary, initialize it with the current score difference
        highest_score_diff_per_season[season] = {
            'score_diff': score_diff,
            'game': index
        }


In [73]:
# Show the game with the highest score difference for each season
for season, result in highest_score_diff_per_season.items():
    game_index = result['game']
    score_diff = result['score_diff']
    print(f"Season: {season}, Game with highest score difference: {game_index}, Score Difference: {score_diff}")

Season: 1985, Game with highest score difference: 236, Score Difference: 60
Season: 1986, Game with highest score difference: 4731, Score Difference: 84
Season: 1987, Game with highest score difference: 8240, Score Difference: 73
Season: 1988, Game with highest score difference: 12046, Score Difference: 68
Season: 1989, Game with highest score difference: 16677, Score Difference: 70
Season: 1990, Game with highest score difference: 19502, Score Difference: 76
Season: 1991, Game with highest score difference: 25161, Score Difference: 68
Season: 1992, Game with highest score difference: 27997, Score Difference: 82
Season: 1993, Game with highest score difference: 33858, Score Difference: 81
Season: 1994, Game with highest score difference: 36404, Score Difference: 69
Season: 1995, Game with highest score difference: 39858, Score Difference: 74
Season: 1996, Game with highest score difference: 44653, Score Difference: 91
Season: 1997, Game with highest score difference: 49033, Score Diffe

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [74]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

Q: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [75]:
games_played_winning = df['Wteam'].value_counts()
games_played_losing = df['Lteam'].value_counts()

games_played = games_played_winning.add(games_played_losing, fill_value=0)

# Show the number of games played by each team
for team, count in games_played.items():
    print(f"Team: {team}, Games Played: {count}")

Team: 1101, Games Played: 76
Team: 1102, Games Played: 840
Team: 1103, Games Played: 910
Team: 1104, Games Played: 975
Team: 1105, Games Played: 447
Team: 1106, Games Played: 855
Team: 1107, Games Played: 512
Team: 1108, Games Played: 866
Team: 1109, Games Played: 169
Team: 1110, Games Played: 910
Team: 1111, Games Played: 876
Team: 1112, Games Played: 981
Team: 1113, Games Played: 937
Team: 1114, Games Played: 910
Team: 1115, Games Played: 520
Team: 1116, Games Played: 980
Team: 1117, Games Played: 891
Team: 1118, Games Played: 27
Team: 1119, Games Played: 868
Team: 1120, Games Played: 936
Team: 1121, Games Played: 175
Team: 1122, Games Played: 903
Team: 1123, Games Played: 906
Team: 1124, Games Played: 906
Team: 1125, Games Played: 506
Team: 1126, Games Played: 906
Team: 1127, Games Played: 442
Team: 1128, Games Played: 106
Team: 1129, Games Played: 907
Team: 1130, Games Played: 951
Team: 1131, Games Played: 945
Team: 1132, Games Played: 895
Team: 1133, Games Played: 949
Team: 1134, 

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [76]:
# We greate a new column with the name "ScoreDifference"
df['ScoreDifference'] = df['Wscore'] - df["Lscore"]

idxmax_per_season = df.groupby('Season')['ScoreDifference'].idxmax()

# Retrieve the corresponding rows using the index and print the results
for season, idx in idxmax_per_season.items():
    game = df.loc[idx]
    print(f"Season: {season}, Game with highest score difference: {idx}")
    print(game)
    print()

Season: 1985, Game with highest score difference: 236
Season             1985
Daynum               33
Wteam              1361
Wscore              128
Lteam              1288
Lscore               68
Wloc                  H
Numot                 0
ScoreDifference      60
HighScoringGame     Yes
Name: 236, dtype: object

Season: 1986, Game with highest score difference: 4731
Season             1986
Daynum               60
Wteam              1314
Wscore              129
Lteam              1264
Lscore               45
Wloc                  N
Numot                 0
ScoreDifference      84
HighScoringGame     Yes
Name: 4731, dtype: object

Season: 1987, Game with highest score difference: 8240
Season             1987
Daynum               51
Wteam              1155
Wscore              112
Lteam              1118
Lscore               39
Wloc                  H
Numot                 0
ScoreDifference      73
HighScoringGame     Yes
Name: 8240, dtype: object

Season: 1988, Game with highest scor

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [77]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [78]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [79]:
type(df['Wscore'])

pandas.core.series.Series

In [80]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [81]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [82]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [83]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y